### PR TITLE
Feature: Add indexation of ERC20 wrapper contracts

### PR DESCRIPTION
**Pull request by josemarinas**
## Description

Add indexation of the underlying contract in wrapped erc20 tokens

Task: [OS-375](https://aragonassociation.atlassian.net/browse/OS-375)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the test network.

[OS-375]: https://aragonassociation.atlassian.net/browse/OS-375?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/subgraph/tests/token/token-voting.test.ts [hunk 1]
```diff
@@ -251,3 +259,31 @@ test('Run TokenVoting (handleVotingSettingsUpdated) mappings with mock event', (
...
+
+  // handle event
+  handleMembershipContractAnnounced(event);
+
+  // assert
+  tokenVotingPlugin.assertEntity();
+  erc20Contract.assertEntity();
+  erc20WrappedContract.assertEntity();
+
+  clearStore();
```
Comment (Rekard0) [inline packages/subgraph/tests/token/token-voting.test.ts:288]: the use of extended classes is abit off.
i think u should not build:
```
erc20Contract.buildOrUpdate();
erc20WrappedContract.buildOrUpdate();
´´´
the handler should have build those, 
then u can have expected changes,
then u can assert

and for the plugin entity u build it with no token,
and then u expect the handler to have updated the plugin entity
then u can assert

File: packages/subgraph/tests/constants.ts [hunk 2]
```diff
@@ -11,6 +11,7 @@ export const ADDRESS_FIVE = '0x0000000000000000000000000000000000000005';
 export const ADDRESS_SIX = '0x0000000000000000000000000000000000000006';
 export const DAO_ADDRESS = '0x00000000000000000000000000000000000000da';
 export const CONTRACT_ADDRESS = '0x00000000000000000000000000000000000000Ad';
+export const WRAPPED_CONTRACT_ADDRESS = '0x0000000000000000000000000000000000000Ad1';
```
Comment (Rekard0) [inline packages/subgraph/tests/constants.ts:14]: u can not use `DAO_TOKEN_ADDRESS`got wrapped token as well?

File: packages/subgraph/tests/token/token-voting.test.ts [hunk 3]
```diff
@@ -251,3 +259,31 @@ test('Run TokenVoting (handleVotingSettingsUpdated) mappings with mock event', (
 
   clearStore();
 });
+
+test('Run TokenVoting (handleMembershipContractAnnounced) with a wrapped token and a mocke event', () => {
```
Comment (Rekard0) [inline packages/subgraph/tests/token/token-voting.test.ts:263]: i would also do more testing,
such as with Governace erc20...etc

File: packages/subgraph/tests/constants.ts (hunk 2 above)
Comment (josemarinas) [reply packages/subgraph/tests/constants.ts:14]: I could use any of them, but I added it for the "readability" of the tests. If you prefer it, I can delete It and use any other address

File: packages/subgraph/src/utils/tokens/erc20.ts [hunk 4]
```diff
@@ -1,11 +1,60 @@
...
 } from '../../../generated/schema';
 import {ERC20} from '../../../generated/templates/DaoTemplateV1_0_0/ERC20';
+import {GovernanceWrappedERC20} from '../../../generated/templates/GovernanceWrappedERC20/GovernanceWrappedERC20';
 import {ERC20_transfer, ERC20_transferFrom, getTransferId} from './common';
+import {supportsInterface} from '../erc165';
+
+export function supportsERC20Wrapped(token: Address): bool {
+  // Double check that it's ERC721 by calling supportsInterface checks.
+  let erc20Wrapped = GovernanceWrappedERC20.bind(token);
+  let introspection_12cdd3ac = supportsInterface(erc20Wrapped, '12cdd3ac'); // GovernanceWrappedERC20
```
Comment (brickpop) [inline packages/subgraph/src/utils/tokens/erc20.ts:16]: why not naming it like "introspectionErc20Wrapped" or similar?

File: packages/subgraph/tests/token/token-voting.test.ts [hunk 5]
```diff
@@ -251,3 +266,87 @@ test('Run TokenVoting (handleVotingSettingsUpdated) mappings with mock event', (
...
+    );
+    // create token calls
+    createTokenCalls(
+      CONTRACT_ADDRESS,
+      'Test Token',
+      'TT',
+      '18',
+      ERC20_AMOUNT_FULL
+    );
+    createWrappedTokenCalls(
```
Comment (Rekard0) [inline packages/subgraph/tests/token/token-voting.test.ts:286]: this should be a mockCall_createWrappedTokenCalls
similar to https://github.com/aragon/osx/blob/develop/packages/subgraph/tests/helpers/method-classes.ts#L160

File: packages/subgraph/tests/token/token-voting.test.ts [hunk 6]
```diff
@@ -251,3 +266,87 @@ test('Run TokenVoting (handleVotingSettingsUpdated) mappings with mock event', (
...
+    getSupportsInterface(CONTRACT_ADDRESS, WRAPPED_ERC20_INTERFACE, false);
+    getSupportsInterface(CONTRACT_ADDRESS, '00000000', false);
+    getSupportsInterface(
+      WRAPPED_CONTRACT_ADDRESS,
+      WRAPPED_ERC20_INTERFACE,
+      true
+    );
+    getSupportsInterface(WRAPPED_CONTRACT_ADDRESS, '00000000', false);
+  });
+  test('it should create an erc20 and assing its address to the tokenVotingPlugin', () => {
```
Comment (Rekard0) [inline packages/subgraph/tests/token/token-voting.test.ts:303]:
```suggestion
  test('it should create an erc20 and assign its address to the tokenVotingPlugin', () => {
```

File: packages/subgraph/tests/constants.ts (hunk 2 above)
Comment (Rekard0) [reply packages/subgraph/tests/constants.ts:14]: u can create `const WRAPPED_CONTRACT_ADDRESS = CONTRACT_ADDRESS` inside your individual test on in the global constants as u wish

File: packages/subgraph/tests/token/token-voting.test.ts [hunk 7]
```diff
@@ -251,3 +266,87 @@ test('Run TokenVoting (handleVotingSettingsUpdated) mappings with mock event', (
...
+    tokenVotingPlugin.token = WRAPPED_CONTRACT_ADDRESS.toLowerCase();
+    tokenVotingPlugin.buildOrUpdate();
+
+    let event = tokenVotingPlugin.createEvent_MembershipContractAnnounced();
+
+    // handle event
+    handleMembershipContractAnnounced(event);
+
+    // assert
+    tokenVotingPlugin.assertEntity();
```
[the remark this answers is not in the document]
Comment (josemarinas) [reply packages/subgraph/tests/token/token-voting.test.ts:346]: Yes the `TokenVotingPlugin` should update Its token field to the address of the WrappingErc20 contract if everything is alright

[the remark this answers is not in the document]
Comment (Rekard0) [reply packages/subgraph/tests/token/token-voting.test.ts:324]: then do assert immediately after buildOrUpdate()
and keep this one as well

[the remark this answers is not in the document]
Comment (josemarinas) [reply packages/subgraph/tests/token/token-voting.test.ts:325]: 👍

File: packages/subgraph/src/utils/tokens/erc20.ts [hunk 8]
```diff
@@ -1,11 +1,61 @@
...
+export function supportsERC20Wrapped(token: Address): bool {
+  // Double check that it's ERC20Wrapped by calling supportsInterface checks.
+  let erc20Wrapped = GovernanceWrappedERC20.bind(token);
+  let introspection_wrapped_erc20 = supportsInterface(erc20Wrapped, WRAPPED_ERC20_INTERFACE); // GovernanceWrappedERC20
+  let introspection_00000000 = supportsInterface(
+    erc20Wrapped,
+    '00000000',
+    false
+  );
+  return introspection_wrapped_erc20 && introspection_00000000;
```
Comment (brickpop) [inline packages/subgraph/src/utils/tokens/erc20.ts:23]: Why don't we simply return `false` as soon as we detect that `introspection_wrapped_erc20` is false? No need to keep computing/calling

File: packages/subgraph/tests/helpers/method-classes.ts [hunk 9]
```diff
@@ -146,6 +151,48 @@ class ERC721TransferMethods extends ERC721Transfer {
...
+  mockCall_createTokenCalls(totalSupply: string | null = null): void {
+    if (!this.name) {
+      throw new Error('Name is null');
+    }
+
+    if (!this.symbol) {
+      throw new Error('Symbol is null');
+    }
+
+    if (!this.underlyingToken) {
```
Comment (brickpop) [inline packages/subgraph/tests/helpers/method-classes.ts:174]:
```suggestion
    } else if (!this.symbol) {
      throw new Error('Symbol is null');
    } else if (!this.underlyingToken) {
```

**Review by brickpop — APPROVED** [review]

File: packages/subgraph/tests/token/token-voting.test.ts [hunk 10]
```diff
@@ -217,3 +227,66 @@ test('Run TokenVoting (handleVotingSettingsUpdated) mappings with mock event', (
 
   clearStore();
 });
+describe('handleMembershipContractAnnounced', () => {
+  beforeAll(() => {});
```
Comment (brickpop) [inline packages/subgraph/tests/token/token-voting.test.ts:231]: Needed?

**Review by Rekard0 — DISMISSED** [review]
LGTM!

File: packages/subgraph/src/utils/tokens/erc20.ts [hunk 11]
```diff
@@ -1,11 +1,67 @@
...
+  ); // GovernanceWrappedERC20
+  if (!introspection_wrapped_erc20) {
+    return false;
+  }
+  let introspection_00000000 = supportsInterface(
+    erc20Wrapped,
+    '00000000',
+    false
+  );
+  return introspection_wrapped_erc20 && introspection_00000000;
```
Comment (brickpop) [inline packages/subgraph/src/utils/tokens/erc20.ts:29]: With the early return we don't need this, now

```suggestion
  return introspection_00000000;
```

**Review by brickpop — APPROVED** [review]

**Review by Rekard0 — APPROVED** [review]
LGTM!

**Pull request merged**